src/shared/sleep-config.c

```diff
@@ -229,6 +245,51 @@ int find_hibernate_location(char **device, char **type, uint64_t *size, uint64_t
                                 continue;
                         }
 
+                        /* Check if swap matches the resume= and resume_offset= values if configured. */
+                        if (!streq(resume_offset, "0") && !streq(resume, "0:0")) {
+                                _cleanup_close_ int fd;
+                                _cleanup_free_ char *swap_device = NULL;
+                                struct stat sb;
+                                int btrfs;
+
+                                fd = open(swap->device, O_RDONLY|O_CLOEXEC|O_NOCTTY);
+                                if (!fd)
+                                        return log_error_errno(errno, "Failed to open %s: %m", swap->device);
+
+                                r = fstat(fd, &sb);
+                                if (r < 0)
+                                        return log_error_errno(errno, "Failed to stat %s: %m", swap->device);
+
+                                if (asprintf(&swap_device, "%u:%u", major(sb.st_dev), minor(sb.st_dev)) < 0)
+                                        return log_oom();
+
+                                btrfs = btrfs_is_filesystem(fd);
+                                if (btrfs < 0)
+                                        return log_error_errno(r, "Error checking %s for Btrfs filesystem: %m", swap->device);
+                                else if (btrfs)
+                                        log_debug("Detection of swap file offset on Btrfs is not supported: %s", swap->device);
+
+                                /* Calculate the swap file offset. */
+                                if (streq(resume, swap_device) && !btrfs) {
+                                        _cleanup_free_ struct fiemap *fiemap = NULL;
+                                        _cleanup_free_ char *offset_str = NULL;
+                                        uint64_t offset;
+
+                                        r = read_fiemap(fd, &fiemap);
+                                        offset = fiemap->fm_extents[0].fe_physical / page_size();
+
+                                        if (asprintf(&offset_str, "%" PRIu64, offset) < 0)
+                                                log_oom();
+
+                                        /* If the swapfile offset matches the resume_offset, use it. */
+                                        if (streq(resume_offset, offset_str)) {
```

Comment:
Please don't do comparisons using strings. Convert "resume=" into a integer and use that for comparisons.